x-pack/plugins/security_solution/server/request_context_factory.ts

```diff
@@ -98,6 +98,8 @@ export class RequestContextFactory implements IRequestContextFactory {
     // time it is requested (see `getEndpointAuthz()` below)
     let endpointAuthz: Immutable<EndpointAuthz>;
 
+    const rulesClient = await startPlugins.alerting.getRulesClientWithRequest(request);
```

Comment:
The other clients in this request context are evaluated lazily, and I don't think we should break that pattern here. You can either revert these changes, or define a shared thunk (a la `getAppClient` in this file).

Reply:
What are your concerns with breaking the pattern here?

Reply:
The concern is that we're now invoking `getRulesClientWithRequest` on every request, whereas previously it was only invoked when we needed it.

Reply:
Sorry for the late response! I see what you mean. In the `RequestContextFactory` you already initiate the following on each request so I followed the same pattern to avoid making massive changes in the security solution code.

```
const [_, startPlugins] = await core.getStartServices();
    const frameworkRequest = await buildFrameworkRequest(context, request);
    const coreContext = await context.core;
    const licensing = await context.licensing;
    const actionsClient = await startPlugins.actions.getActionsClientWithRequest(request);

    const dataViewsService = await startPlugins.dataViews.dataViewsServiceFactory(
      coreContext.savedObjects.client,
      coreContext.elasticsearch.client.asInternalUser,
      request
    );
```

You are right, ideally, we should not do async calls in the `create` method of the `RequestContextFactory`. I noticed that the rules client is being used in `getDetectionRulesClient` and `getDetectionEngineHealthClient`. I can move the `await startPlugins.alerting.getRulesClientWithRequest(request);` inside the above functions but they have to become `async` (now they are not).

Reply:
I noticed that the context is not created on each request but only when you want access to the specific context. For example

```
const caseContext = await context.cases;
const casesClient = await caseContext.getCasesClient();
```

Reply:
Fair enough. Since fetching the rules client is now async, and we already `await` other clients on every request as you said, these changes are likely inconsequential and the best path forward (we definitely don't want to be changing the signature of these common context methods).

TL;DR I'm fine with this as is 👍